pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,219 @@
+// and calculate the flush period based on actual timeouts, with a
+// 10 minutes default
+func (a *TokenTimeoutUpdater) updateClientTimeouts() time.Duration {
+	// Default to 10 minutes
+	var flush int32 = 600
+
+	// Update known ones only
+	if len(a.clientTimeouts) > 0 {
+		for name := range a.clientTimeouts {
+			c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+			if err != nil {
+				if kerrors.IsNotFound(err) {
+					delete(a.clientTimeouts, name)
+				}
+			}
+			if c.AccessTokenTimeoutSeconds == nil {
+				delete(a.clientTimeouts, name)
+			} else {
+				a.clientTimeouts[name] = *c.AccessTokenTimeoutSeconds
+				// Use half period if it is smaller
+				half := int32(math.Ceil(float64(a.clientTimeouts[name]) / 2))
+				if flush > half {
+					flush = half
+				}
+			}
+		}
+	}
+	return time.Duration(flush) * time.Second
+}
+
+func (a *TokenTimeoutUpdater) clientTimeout(name string) int32 {
+	t, ok := a.clientTimeouts[name]
+	if !ok {
+		c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				glog.Info("Token timeout set but OAuthClient %s not found!", name)
+			} else {
+				glog.Error("Failed to fetch OAuthClient for timeout value")
+			}
+			return 0
+		}
+		if c.AccessTokenTimeoutSeconds == nil {
+			glog.Info("Token timeout set but OAuthClient %s has no timeout!", name)
+			return 0
+		}
+		t = *c.AccessTokenTimeoutSeconds
+		a.clientTimeouts[name] = t
+	}
+	return t
+}
+
+func (a *TokenTimeoutUpdater) insert(td tokenData) {
+	a.data[td.name] = td
+	a.tree.ReplaceOrInsert(&tokenDataRef{td.name, td.timeout})
+}
+
+func (a *TokenTimeoutUpdater) remove(td tokenData) {
+	a.tree.Delete(&tokenDataRef{td.name, td.timeout})
+	delete(a.data, td.name)
+}
+
+func (a *TokenTimeoutUpdater) flush(flushHorizon time.Time) {
+	flushedTokens := 0
+	totalTokens := len(a.data)
+
+	glog.Infof("Flushing tokens timing out before %v", flushHorizon)
+
+	for {
+		item := a.tree.Min()
+		if item == nil {
+			// out of items
+			break
+		}
+		td := a.data[item.(*tokenDataRef).name]
+		if td.timeout.Before(flushHorizon) {
+			delta := a.clientTimeout(td.client)
+			// ignore tokens for which we can't find a delta
+			if delta > 0 {
+				token, err := a.tokens.Get(td.name, metav1.GetOptions{})
+				if err == nil { // just ignore if the token is gone
+					newtimeout := int32(td.seen.Sub(token.CreationTimestamp.Time)/time.Second) + delta
+					if newtimeout > token.TimeoutsIn {
+						token.TimeoutsIn = newtimeout
+						// ignore errors, if it fails to update there isn't much we can do
+						_, _ = a.tokens.Update(token)
+					}
+				}
+			}
+			a.remove(td)
+			flushedTokens++
+		} else {
+			// out of items within the flush Horizon
+			break
+		}
+	}
+
+	glog.Infof("Flushed %d tokens out of %d in bucket", flushedTokens, totalTokens)
+}
+
+func (a *TokenTimeoutUpdater) Start() {
+	// TODO: make safety margin configurable ?
+	var safetyMargin time.Duration = time.Second * 15
```

Comment:
Seems a bit small?

Reply:
Note that 15 seconds is above the current timeout, not the delta we consider.
Do you think a single update take more than 15 seconds ?

Reply:
I have seen two minute update times in online, but I am not sure we care about such cases.

Reply:
If an update takes so long than timeouts will already fire left and right.
We can tweak this value if we have some data, I was thinking for example of making it a percentage of the delta instead of a fixed amount (with a minimum) would that make sense ?